### PR TITLE
cmd/tailscale: write fewer known_hosts, resolve ssh host to FQDN early

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #3802

### CODE REVIEW

**Review by maisem — COMMENTED** [review]
~Won't this only work if magic dns is enabled?~ I misunderstood what this was doing.

File: cmd/tailscale/cli/ssh.go [hunk 1]
```diff
@@ -138,25 +149,40 @@ func genKnownHosts(st *ipnstate.Status) []byte {
...
+			return dnsName, true
 		}
-		if base, _, ok := strings.Cut(ps.DNSName, "."); ok {
-			addEntries(base)
+		if base, _, ok := strings.Cut(ps.DNSName, "."); ok && strings.EqualFold(base, arg) {
+			return dnsName, true
 		}
-		for _, ip := range st.TailscaleIPs {
-			addEntries(ip.String())
+		for _, ip := range ps.TailscaleIPs {
```
Comment (maisem) [inline cmd/tailscale/cli/ssh.go:181]: reorder to bring this to the top? you can save a bunch of cuts and string comparisons. not that it matters much here :) 
and you can conditionally only look at IPs if `!argIP.IsZero()`

File: cmd/tailscale/cli/ssh.go [hunk 2]
```diff
@@ -138,25 +149,40 @@ func genKnownHosts(st *ipnstate.Status) []byte {
...
+			fmt.Fprintf(&buf, "%s %s\n", host, hostKey)
 		}
-		if ps.DNSName != "" {
-			addEntries(ps.DNSName)
+	}
+	return buf.Bytes()
+}
+
+// nodeDNSNameFromArg returns the PeerStatus.DNSName value (but
+// without the trailing dot) from a peer in st that matches the input
```
Comment (maisem) [inline cmd/tailscale/cli/ssh.go:166]: keep the trailing dot, no reason not to.

File: cmd/tailscale/cli/ssh.go [hunk 3]
```diff
@@ -138,25 +149,40 @@ func genKnownHosts(st *ipnstate.Status) []byte {
...
+// arg (which can be a base name, full DNS name, or an IP)
```
Comment (maisem) [inline cmd/tailscale/cli/ssh.go:167]: missing period at the end

**Review by maisem — APPROVED** [review]

File: cmd/tailscale/cli/ssh.go [hunk 4]
```diff
@@ -135,28 +146,40 @@ func genKnownHosts(st *ipnstate.Status) []byte {
...
+	argIP, _ := netaddr.ParseIP(arg)
+	for _, ps := range st.Peer {
+		dnsName = ps.DNSName
+		if !argIP.IsZero() {
+			for _, ip := range ps.TailscaleIPs {
+				if ip == argIP {
+					return dnsName, true
 				}
-				fmt.Fprintf(&buf, "%s %s\n", host, hostKey)
 			}
```
Comment (maisem) [inline cmd/tailscale/cli/ssh.go:175]: continue here. can't possibly match anything else

**Review by maisem — APPROVED** [review]

File: cmd/tailscale/cli/ssh.go (hunk 4 above)
Comment (bradfitz) [reply cmd/tailscale/cli/ssh.go:175]: Done

**Review by maisem — APPROVED** [review]

**Pull request merged**